### PR TITLE
feat: support heartbeat mechanism between ts-sql and ts-meta

### LINKED ISSUE #308: support task assignment and status report of cq
### Is your feature request related to a problem? Please describe.（如果该特性或者功能和已知某个问题相关，请描述问题是什么）

1. ts-sql needs to periodically fetch cqs from ts-meta to execute or stop cq.
2. ts-sql needs to periodically send heartbeats to ts-meta to prove its survival.
3. The last run time of cq needs to be sent to ts-meta for persistence.

### Describe the solution you'd like（问题的解决办法是什么）

The existing designs are as follows:
1. ts-sql periodically sends heartbeats to ts-meta, reports its own survival.
2. ts-meta records the relevant information of ts-sql according to the heartbeat of ts-sql, but does not need to be persisted.
3. In the meta data of CQ, mark the ts-sql of the CQ that should be executed, so that ts-sql knows which CQs need to be executed and which ones need to be stopped after getting the meta data.
4. When ts-sql receives the task of executing cq, it starts a goroutine to execute cq, and after completing a `select into` operation, reports the last run time to ts-meta for persistence.
5. When ts-sql receives the task of stopping cq, it closes the corresponding goroutine.
6. Store all cqs running on each ts-sql in ts-meta for scheduling and load balancing.

### Describe alternatives you've considered（描述您考虑过的替代方案）

No alternatives.

### Additional context（其他补充内容）

The scheduling and load balancing of cq can be further improved on this basis.

### PR DESCRIPTION
**Pull request by Benevor**
### What problem does this PR solve?

Issue Number: ref #308

### What is changed and how it works?
Implement the sending and processing of `Sql2MetaHeartbeat` RPC message . Besides, the ts-sql downtime detection mechanism is implemented.

### How Has This Been Tested?

In the corresponding test file, I added test functions for the `Sql2MetaHeartbeat` message. In addition, I verified the correctness of the logic of the program by manually debugging the program and printing logs.
- [x] full build of project
- [x] service_test.go
- [x] processor_test.go
- [x] handlers_process_test.go
- [x] meta_callback_test.go

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Review by shilinlee — CHANGES_REQUESTED** [review]

File: app/ts-sql/sql/server.go [hunk 1]
```diff
@@ -112,7 +112,7 @@ func NewServer(conf config.Config, cmd *cobra.Command, logger *Logger.Logger) (a
 		cmd:         cmd,
 		Logger:      logger,
 		httpService: httpd.NewService(c.HTTP),
-		cqService:   continuousquery.NewService(c.ContinuousQuery.RunInterval),
+		cqService:   continuousquery.NewService(c.ContinuousQuery.RunInterval, c.HTTP.BindAddress),
```
[the remark this answers is not in the document]
Comment (Benevor) [reply app/ts-sql/sql/server.go]: I have fixed it.

File: app/ts-meta/meta/store_fsm.go [hunk 2]
```diff
@@ -200,6 +200,10 @@ func (fsm *storeFSM) executeCmd(cmd proto2.Command) interface{} {
 		return fsm.applyUpdatePtVersionCommand(&cmd)
 	case proto2.Command_CreateContinuousQueryCommand:
 		return fsm.applyCreateContinuousQueryCommand(&cmd)
+	case proto2.Command_HeartbeatCommand:
+		return fsm.applyHeartbeatCommand(&cmd)
```
[the remark this answers is not in the document]
Comment (Benevor) [reply app/ts-meta/meta/store_fsm.go]: The sending and processing of heartbeat has been separated from the logic of raft.

File: app/ts-meta/meta/store.go [hunk 3]
```diff
@@ -1374,6 +1387,123 @@ func (s *Store) getMeasurementInfo(dbName, rpName, mstName string) ([]byte, erro
...
+			}
+		}
+	}
+	sqli.Cqs = nil
+	return nil
+}
+
+func (s *Store) detectTSSqlCrash() {
+	defer s.wg.Done()
+	checkTime := time.After(detectTSSqlCrashInterval)
```
Comment (shilinlee) [inline app/ts-meta/meta/store.go:1448]: use `time.Ticker`?

File: config/openGemini.singlenode.conf [hunk 4]
```diff
@@ -9,6 +9,7 @@
 
 [http]
   bind-address = "127.0.0.1:8086"
+  domain = "localhost"
```
Comment (shilinlee) [inline config/openGemini.singlenode.conf:12]: `domain` is only used in cloud-native scene

File: lib/metaclient/meta_client.go [hunk 5]
```diff
@@ -683,6 +684,63 @@ func (c *Client) GetMeasurementInfoStore(dbName string, rpName string, mstName s
 	return mst, nil
 }
 
+func (c *Client) RetryHeartbeat(id string, cqs []string) error {
```
Comment (shilinlee) [inline lib/metaclient/meta_client.go:687]: Can we use `retryExec` to replace with `RetryHeartbeat `

**Review by shilinlee — CHANGES_REQUESTED** [review]

**Review by shilinlee — CHANGES_REQUESTED** [review]

File: app/ts-meta/meta/handlers.gen.go [hunk 6]
```diff
@@ -303,3 +309,22 @@ func (h *GetMeasurementInfo) SetRequestMsg(data transport.Codec) error {
 func (h *GetMeasurementInfo) Instance() RPCHandler {
 	return &GetMeasurementInfo{}
 }
+
+type Heartbeat struct {
```
[the remark this answers is not in the document]
Comment (Benevor) [reply app/ts-meta/meta/handlers.gen.go]: I have fixed.

File: app/ts-meta/meta/store.go (hunk 3 above)
Comment (Benevor) [reply app/ts-meta/meta/store.go]: I have fixed.

File: config/openGemini.singlenode.conf (hunk 4 above)
Comment (Benevor) [reply config/openGemini.singlenode.conf]: I have fixed.

File: lib/metaclient/meta_client.go (hunk 5 above)
Comment (Benevor) [reply lib/metaclient/meta_client.go]: `retryExec` only send `ExecuteRequestMessage`.

File: app/ts-meta/meta/message/message.go [hunk 7]
```diff
@@ -209,3 +209,12 @@ type GetMeasurementInfoResponse struct {
 	Data []byte
 	Err  string
 }
+
+type HeartbeatRequest struct {
+	Id  string
```
[the remark this answers is not in the document]
Comment (Benevor) [reply app/ts-meta/meta/message/message.go]: I have fixed.

File: app/ts-meta/meta/tssqlinfo.go [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+
+package meta
+
+import (
+	"container/list"
+	"time"
+)
+
+// TSSQLInfo represents metadata about a ts-sql.
+type TSSQLInfo struct {
```
[the remark this answers is not in the document]
Comment (Benevor) [reply app/ts-meta/meta/tssqlinfo.go]: I have fixed.

File: app/ts-meta/meta/store.go [hunk 9]
```diff
@@ -63,6 +64,11 @@ const (
 	updateCacheInterval = 100 * time.Millisecond
 
 	pushInterval = 10 * time.Second
+
+	detectTSSqlCrashInterval = time.Minute
+
+	// HeartbeatTimeoutDuration represents ts-sql heartbeat timeout
+	HeartbeatTimeoutDuration = time.Minute
```
[the remark this answers is not in the document]
Comment (Benevor) [reply app/ts-meta/meta/store.go]: I have fixed.

File: app/ts-sql/sql/server.go [hunk 10]
```diff
@@ -112,7 +112,7 @@ func NewServer(conf config.Config, cmd *cobra.Command, logger *Logger.Logger) (a
 		cmd:         cmd,
 		Logger:      logger,
 		httpService: httpd.NewService(c.HTTP),
-		cqService:   continuousquery.NewService(c.ContinuousQuery.RunInterval),
+		cqService:   continuousquery.NewService(c.ContinuousQuery.RunInterval, config.CombineDomain(c.HTTP.Domain, c.HTTP.BindAddress)),
```
[the remark this answers is not in the document]
Comment (Benevor) [reply app/ts-sql/sql/server.go]: I use `c.HTTP.BindAddress` as ts-sql `host`. But cq service do not need a `host` now.

File: app/ts-meta/meta/store.go [hunk 11]
```diff
@@ -63,6 +64,11 @@ const (
 	updateCacheInterval = 100 * time.Millisecond
 
 	pushInterval = 10 * time.Second
+
+	detectTSSqlCrashInterval = time.Minute
```
[the remark this answers is not in the document]
Comment (Benevor) [reply app/ts-meta/meta/store.go]: OK.

**Review by shilinlee — CHANGES_REQUESTED** [review]

File: app/ts-meta/meta/store.go [hunk 12]
```diff
@@ -1374,6 +1387,84 @@ func (s *Store) getMeasurementInfo(dbName, rpName, mstName string) ([]byte, erro
...
+	if err := s.updateSqlNode(host, ts); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (s *Store) updateSqlNode(host string, ts time.Time) error {
+	// update HeartbeatInfoList.
+	if s.SqlNodes[host] != nil {
```
[the remark this answers is not in the document]
Comment (Benevor) [reply app/ts-meta/meta/store.go:1407]: `updateSqlNode()` only called by `applySql2MetaHeartbeat()`.  `applySql2MetaHeartbeat()` will deal with the concurrency race issues.

File: app/ts-meta/meta/store.go [hunk 13]
```diff
@@ -1374,6 +1387,84 @@ func (s *Store) getMeasurementInfo(dbName, rpName, mstName string) ([]byte, erro
...
+func (s *Store) detectSqlNodeCrash() {
+	defer s.wg.Done()
+	ticker := time.NewTicker(detectTSSqlCrashInterval)
+	for {
+		select {
+		case <-s.closing:
+			return
+		case <-ticker.C:
+			if err := s.detectCrash(); err != nil {
+				s.Logger.Error("detect ts-sql whether crashed or not, error", zap.Error(err))
```
[the remark this answers is not in the document]
Comment (Benevor) [reply app/ts-meta/meta/store.go]: I have fixed it.

File: open_src/influx/meta/proto/meta.proto [hunk 14]
```diff
@@ -294,6 +294,7 @@ message Command {
 		ExpandGroupsCommand                        = 81;
 		UpdatePtVersionCommand                     = 82;
 		CreateContinuousQueryCommand               = 83;
+		Sql2MetaHeartbeatCommand				   = 84;
```
[the remark this answers is not in the document]
Comment (Benevor) [reply open_src/influx/meta/proto/meta.proto]: I have fixed it.

**Review by shilinlee — APPROVED** [review]

**Pull request merged**